File: ios/Processors/EnrollmentProcessor.swift
import AVFoundation

// The same as EnrollmentProcessor.java
// Instead of the demo app and SDK requirements it implements SessionDelegate not FaceTecFaceScanProcessorDelegate
// To proxy callbacks to FaceTec we create instance of SessionProcessingDelegate
// This twisted was is needed to avoid crash during build
class EnrollmentProcessor: NSObject, URLSessionTaskDelegate, SessionDelegate {
    var maxRetries: Int?
    var enrollmentIdentifier: String!
    private let defaultMaxRetries = -1

    var isSuccess = false
    var lastMessage: String!
    var lastResult: FaceTecSessionResult!
    var resultCallback: FaceTecFaceScanResultCallback!
    var retryAttempt = 0
    var timeout: TimeInterval? = nil

    var processingDelegate: ProcessingDelegate
    var presentSessionVCFrom: UIViewController
    var sessionDelegate: SessionProcessingDelegate?

    private var alwaysRetry: Bool {
        get {
            return maxRetries == nil || maxRetries! < 0
        }
    }

    init(fromVC: UIViewController, delegate: ProcessingDelegate) {
        self.processingDelegate = delegate
        self.presentSessionVCFrom = fromVC
      
        super.init()
        // instantiating a session delegate ObjectiveC class
        // wrapping the enrollment processor
        self.sessionDelegate = SessionProcessingDelegate(session: self)
    }

    func enroll(_ enrollmentIdentifier: String, _ maxRetries: Int? = nil, _ timeout: Int? = nil) {
        requestCameraPermissions() {
            self.startSession() { sessionToken in
                self.enrollmentIdentifier = enrollmentIdentifier
                self.maxRetries = self.defaultMaxRetries

                // there're issues with passing nil / null for numbers
                // we're passing -1 if no param was set on JS side
                // so here we have to add additional > (or >=) 0 check
                if maxRetries != nil && maxRetries! >= 0 {
                    self.maxRetries = maxRetries!
                }

                if timeout != nil && timeout! > 0 {
                  self.timeout = Double(timeout!) / 1000
                }

                DispatchQueue.main.async {
                    // the sessionDelegate now passes to createSessionVC instead of EnrollmentProcessor
                    let sessionVC = FaceTec.sdk.createSessionVC(faceScanProcessorDelegate: self.sessionDelegate!, sessionToken: sessionToken)

                    self.presentSessionVCFrom.present(sessionVC, animated: true, completion: {
                        EventEmitter.shared.dispatch(.UI_READY)
                    })
                }
            }
        }
    }

    func onFaceTecSessionResult(sessionResult: FaceTecSessionResult, faceScanResultCallback: FaceTecFaceScanResultCallback) {
        lastResult = sessionResult
        resultCallback = faceScanResultCallback

        if sessionResult.status != .sessionCompletedSuccessfully {
            FaceVerification.shared.cancelInFlightRequests()
            faceScanResultCallback.onFaceScanResultCancel()
            return
        }


        // notifying that capturing is done
        EventEmitter.shared.dispatch(.CAPTURE_DONE)

        // perform verification
        sendEnrollmentRequest()
    }

    func onFaceTecSessionDone() {
        processingDelegate.onProcessingComplete(isSuccess: isSuccess, sessionResult: lastResult, sessionMessage: lastMessage)
    }

    // upload progress handler, requested by URLSessionTaskDelegate
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
        // get progress while performing the upload
        let uploaded: Float = Float(totalBytesSent) / Float(totalBytesExpectedToSend)

        // updating the UX, upload progress from 10 to 80%
        resultCallback.onFaceScanUploadProgress(uploadedPercent: 0.1 + 0.7 * uploaded)

        if (totalBytesSent == totalBytesExpectedToSend) {
            let processingMessage = NSMutableAttributedString.init(string: Customization.resultFacescanProcessingMessage)

            // switch status message to processing once upload completed
            resultCallback.onFaceScanUploadMessageOverride(uploadMessageOverride: processingMessage)
        }
    }

    private func startSession(sessionTokenCallback: @escaping (String) -> Void) {
        FaceVerification.shared.getSessionToken() { sessionToken, error in
            if error != nil {
                self.processingDelegate.onSessionTokenError()
                return
            }

            sessionTokenCallback(sessionToken!)
        }
    }

    private func requestCameraPermissions(requestCallback: @escaping () -> Void) {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            requestCallback()
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { (granted) in
                if (granted) {
                    requestCallback()
                } else {
                    self.processingDelegate.onCameraAccessError()
                }
            }
        case .denied,
             .restricted:
            processingDelegate.onCameraAccessError()
        @unknown default:
            processingDelegate.onCameraAccessError()
        }
    }

    private func sendEnrollmentRequest() -> Void {
        // setting initial progress to 0 for freeze progress bar
        resultCallback.onFaceScanUploadProgress(uploadedPercent: 0)

        let payload: [String : Any] = [
            "faceScan": lastResult.faceScanBase64!,
            "auditTrailImage": lastResult.auditTrailCompressedBase64!.first!,
            "lowQualityAuditTrailImage": lastResult.lowQualityAuditTrailCompressedBase64!.first!,
            "sessionId": lastResult.sessionId
        ]

        FaceVerification.shared.enroll(
            enrollmentIdentifier!,
            payload,
            withTimeout: timeout,
            withDelegate: self
        ) { response, error in
            let enrollmentResult = response?["enrollmentResult"] as? [String: String]
            let resultBlob = enrollmentResult?["resultBlob"]
            var enrollmentError = error
          
            self.resultCallback.onFaceScanUploadProgress(uploadedPercent: 1)
          
            if error == nil && resultBlob == nil {
              enrollmentError = FaceVerificationError.unexpectedResponse
            }

            if enrollmentError != nil {
                self.handleEnrollmentError(enrollmentError!, response)
                return
            }

            self.isSuccess = true
            self.lastMessage = Customization.resultSuccessMessage

            FaceTecCustomization.setOverrideResultScreenSuccessMessage(self.lastMessage)
            self.resultCallback.onFaceScanGoToNextStep(scanResultBlob: resultBlob!)
        }
    }

    private func handleEnrollmentError(_ error: Error, _ response: [String: AnyObject]?) -> Void {
        let faceTecError = (error as? FaceVerificationError) ?? .unexpectedResponse

        // by default we'll use exception's message as lastMessage
        lastMessage = faceTecError.message

        if response != nil {
            let flags = response?["enrollmentResult"] as? [String: Bool]
            let enrollmentResult = response?["enrollmentResult"] as? [String: String]

            // if isDuplicate is strictly true, that means we have dup face
            let isDuplicateIssue = true == flags?["isDuplicate"]
            let is3DMatchIssue = true == flags?["isNotMatch"]
            let isLivenessIssue = false == flags?["isLive"]
            let isEnrolled = true == flags?["isEnrolled"]
            let resultBlob = enrollmentResult?["resultBlob"]

            // if there's no duplicate / 3d match issues but we have
            // liveness issue strictly - we'll check for possible session retry
            if resultBlob != nil && isLivenessIssue && !isDuplicateIssue && !is3DMatchIssue {
                // if haven't reached retries threshold or max retries is disabled
                // (is null or < 0) we'll ask to retry capturing
                if alwaysRetry || retryAttempt < maxRetries! {
                    let retryMessage = NSMutableAttributedString.init(string: lastMessage)

                    // increasing retry attempts counter
                    retryAttempt += 1
                    // showing reason
                    resultCallback.onFaceScanUploadMessageOverride(uploadMessageOverride: retryMessage)
                    // notifying about retry
                    resultCallback.onFaceScanGoToNextStep(scanResultBlob: resultBlob!)

                    // dispatching retry event
                    EventEmitter.shared.dispatch(.FV_RETRY, [
                        "reason": lastMessage!,
                        "match3d": !is3DMatchIssue,
                        "liveness": !isLivenessIssue,
                        "duplicate": isDuplicateIssue,
                        "enrolled": isEnrolled
                    ])

                    return
                }
            }
        }

        resultCallback.onFaceScanResultCancel()
    }
}

File: android/src/main/java/org/gooddollar/facetec/processors/EnrollmentProcessor.java
package org.gooddollar.facetec.processors;

import androidx.annotation.Nullable;
import android.content.Context;
import org.json.JSONObject;

import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;

import com.facetec.sdk.FaceTecSDK;
import com.facetec.sdk.FaceTecFaceScanProcessor;
import com.facetec.sdk.FaceTecFaceScanResultCallback;
import com.facetec.sdk.FaceTecSessionActivity;
import com.facetec.sdk.FaceTecSessionStatus;
import com.facetec.sdk.FaceTecSessionResult;
import com.facetec.sdk.FaceTecCustomization;

import org.gooddollar.facetec.api.FaceVerification;
import org.gooddollar.facetec.api.NetworkingHelpers;
import org.gooddollar.facetec.api.ProgressRequestBody;
import okhttp3.RequestBody;

import org.gooddollar.facetec.processors.ProcessingSubscriber;
import org.gooddollar.facetec.util.EventEmitter;
import org.gooddollar.facetec.util.Customization;
import org.gooddollar.facetec.util.Permissions;

// Implements face verification flow. Based on the class from the FaceTec demo app
public class EnrollmentProcessor implements FaceTecFaceScanProcessor {
  private Context context;
  private ProcessingSubscriber subscriber;
  private Permissions permissions;

  private FaceTecFaceScanResultCallback lastResultCallback = null;
  private FaceTecSessionResult lastResult = null;
  private String lastMessage = null;

  private Integer timeout = null;
  private int maxRetries = -1;
  private int retryAttempt = 0;
  private String enrollmentIdentifier = null;
  private String v1Identifier = null;
  private String chainId = null;
  private boolean isSuccess = false;

  // instantiated with content context (activity) and subscriber  
  public EnrollmentProcessor(Context context, ProcessingSubscriber subscriber) {
    this.context = context;
    this.subscriber = subscriber;
    permissions = new Permissions(context);
  }

  public ProcessingSubscriber getSubscriber() {
    return subscriber;
  }

  // different enroll() overloads covering params defaults
  public void enroll(final String enrollmentIdentifier, final String v1Identifier) {
    enroll(enrollmentIdentifier, v1Identifier, null, null, null);
  }
  
  public void enroll(final String enrollmentIdentifier, final String v1Identifier, final String chainId) {
    enroll(enrollmentIdentifier, v1Identifier, chainId, null, null);
  }

  public void enroll(final String enrollmentIdentifier, final String v1Identifier, final String chainId, final Integer maxRetries) {
    enroll(enrollmentIdentifier, v1Identifier, chainId, maxRetries, null);
  }

  // starts FV session
  public void enroll(final String enrollmentIdentifier, final String v1Identifier, @Nullable final String chainId, @Nullable final Integer maxRetries, @Nullable final Integer timeout) {
    final Context ctx = this.context;
    final ProcessingSubscriber subscriber = this.subscriber;

    // get session token callback
    final FaceVerification.SessionTokenCallback onSessionTokenRetrieved =
      new FaceVerification.SessionTokenCallback() {
        @Override
        public void onSessionTokenReceived(String sessionToken) {
          // when got token successfully - show FV UI
          FaceTecSessionActivity.createAndLaunchSession(ctx, EnrollmentProcessor.this, sessionToken);
          EventEmitter.dispatch(EventEmitter.UXEvent.UI_READY);
        }

        @Override
        public void onFailure(FaceVerification.APIException exception) {
          // otherwise reject with specific error
          subscriber.onSessionTokenError(exception);
        }
      };

    // store enrollmentIdentifier, maxRetries and timeout in the corresponding instance vars
    this.enrollmentIdentifier = enrollmentIdentifier;
    this.v1Identifier = v1Identifier;
    this.chainId = chainId;

    if ((maxRetries != null) && (maxRetries >= 0)) {
      this.maxRetries = maxRetries;
    }

    if ((timeout != null) && (timeout > 0)) {
      this.timeout = timeout;
    }

    // request camera permissions.
    this.permissions.requestCameraPermissions(new Permissions.PermissionsCallback() {
      @Override
      public void onSuccess() {
        // on premissions granted - get session token, pas callbacks
        FaceVerification.getSessionToken(onSessionTokenRetrieved);
      }

      @Override
      public void onFailure() {
        // otherwise reject with specific error
        subscriber.onCameraAccessError();
      }
    });
  }

  // capturing done/failed callback
  // logic the same as on the web
  public void processSessionWhileFaceTecSDKWaits(
    final FaceTecSessionResult sessionResult,
    final FaceTecFaceScanResultCallback faceScanResultCallback
  ) {
    lastResult = sessionResult;
    lastResultCallback = faceScanResultCallback;

    if (sessionResult.getStatus() != FaceTecSessionStatus.SESSION_COMPLETED_SUCCESSFULLY) {
      NetworkingHelpers.cancelPendingRequests();
      faceScanResultCallback.cancel();

      return;
    }

    // notifying that capturing is done
    EventEmitter.dispatch(EventEmitter.UXEvent.CAPTURE_DONE);

    // perform verification
    sendEnrollmentRequest();
  }

  // done callback
  // logic the same as on the web
  public void onFaceTecSDKCompletelyDone() {
    // recalls processing subscriber with success slate and last result/message
    subscriber.onProcessingComplete(isSuccess, lastResult, lastMessage);
  }

  // enrollment request factory helper
  private RequestBody createEnrollmentRequest(JSONObject payload) {
    final FaceTecFaceScanResultCallback resultCallback = lastResultCallback;

    // create request with send progress listener
    return new ProgressRequestBody(FaceVerification.jsonStringify(payload),
      new ProgressRequestBody.Listener() {
        // send from total listener, the logic same as on web
        @Override
        public void onUploadProgressChanged(long bytesWritten, long totalBytes) {
          // get progress while performing the upload
          final float uploaded = ((float) bytesWritten) / ((float) totalBytes);

          // updating the UX, upload progress from 10 to 80%
          resultCallback.uploadProgress(0.1f + 0.7f * uploaded);

          if (bytesWritten == totalBytes) {
            // switch status message to processing once upload completed
            resultCallback.uploadMessageOverride(Customization.resultFacescanProcessingMessage);
          }
        }
      }
    );
  }

  // send request helper, processing logic same as on web
  private void sendEnrollmentRequest() {
    final FaceTecFaceScanResultCallback resultCallback = lastResultCallback;
    JSONObject payload = new JSONObject();

    // setting initial progress to 0 for freeze progress bar
    resultCallback.uploadProgress(0);

    try {
      // same request as on web
      // { faceScan, auditTrailImage, lowQualityAuditTrailImage, sessionId, fvSigner }
      payload.put("faceScan", lastResult.getFaceScanBase64());
      payload.put("auditTrailImage", lastResult.getAuditTrailCompressedBase64()[0]);
      payload.put("lowQualityAuditTrailImage", lastResult.getLowQualityAuditTrailCompressedBase64()[0]);
      payload.put("sessionId", lastResult.getSessionId());
      payload.put("fvSigner", this.v1Identifier);

      // if no chainId then DO NOT send chainId in body
      if (this.chainId != null) {
        payload.put("chainId", this.chainId);
      }
    } catch(Exception e) {
      lastMessage = "Exception raised while attempting to create JSON payload for upload.";
      resultCallback.cancel();
    }

    RequestBody request = createEnrollmentRequest(payload);
    FaceVerification.enroll(enrollmentIdentifier, request, timeout, new FaceVerification.APICallback() {
      @Override
      public void onSuccess(JSONObject response) { // same logic as on web
        String successMessage = Customization.resultSuccessMessage;
        JSONObject enrollmentResult = getEnrollmentResult(response);
        String resultBlob = enrollmentResult.optString("resultBlob"); // response.data.resultBlob

        resultCallback.uploadProgress(1);

        // no result blob - throw unknown error
        if (resultBlob == null) {
          FaceVerification.APIException exception = new FaceVerification.APIException(
            FaceVerification.unexpectedMessage, response
          );

          EnrollmentProcessor.this.handleEnrollmentError(exception);
          return;
        }

        // GoodServer returs only success blob
        // any specific error as dup or low quality 
        // throws exception and is processed at onFailure
        EnrollmentProcessor.this.isSuccess = true; // set success sate & message
        EnrollmentProcessor.this.lastMessage = successMessage;
        FaceTecCustomization.overrideResultScreenSuccessMessage = successMessage; // show unicorn greeting at UI

        // finish flow with OK
        resultCallback.succeed();
        resultCallback.proceedToNextStep(resultBlob);
      }

      @Override
      public void onFailure(FaceVerification.APIException exception) {
        resultCallback.uploadProgress(1); // on any error set procressbar complete
        EnrollmentProcessor.this.handleEnrollmentError(exception); // and handle error
      }
    });
  }

  // handles enrollment error. logic the same as on web
  private void handleEnrollmentError(FaceVerification.APIException exception) {
    JSONObject response = exception.getResponse();

    // by default we'll use exception's message as lastMessage
    lastMessage = exception.getMessage();

    if (response != null) {
      JSONObject enrollmentResult = getEnrollmentResult(response);

      // if isDuplicate is strictly true, that means we have dup face
      boolean isDuplicateIssue = enrollmentResult.optBoolean("isDuplicate", false);
      boolean is3DMatchIssue = enrollmentResult.optBoolean("isNotMatch", false);
      boolean isEnrolled = enrollmentResult.optBoolean("isEnrolled", false);
      // in JS code we're checking for false === isLive strictly. so if no isLive flag in the response,
      // we assume that liveness check was successfull. That's why we're setting true as fallback value
      boolean isLivenessIssue = enrollmentResult.optBoolean("isLive", true);
      // getting result Blob to use in the retry case
      String resultBlob = enrollmentResult.optString("resultBlob");

      // if there's no duplicate / 3d match issues but we have
      // liveness issue strictly - we'll check for possible session retry
      if (isLivenessIssue && (resultBlob != null) && !isDuplicateIssue && !is3DMatchIssue) {
        // if haven't reached retries threshold or max retries is disabled
        // (is null or < 0) we'll ask to retry capturing
        if ((maxRetries < 0) || (retryAttempt < maxRetries)) {
          // increasing retry attempts counter
          retryAttempt += 1;
          // showing reason
          lastResultCallback.uploadMessageOverride(lastMessage);
          // notifying about retry
          lastResultCallback.proceedToNextStep(resultBlob);

          // dispatching retry event
          WritableMap eventData = Arguments.createMap();

          eventData.putString("reason", lastMessage);
          eventData.putBoolean("match3d", !is3DMatchIssue);
          eventData.putBoolean("liveness", !isLivenessIssue);
          eventData.putBoolean("duplicate", isDuplicateIssue);
          eventData.putBoolean("enrolled", isEnrolled);

          EventEmitter.dispatch(EventEmitter.UXEvent.FV_RETRY, eventData);
          return;
        }
      }
    }

    lastResultCallback.cancel();
  }

  //  reads response.data.enrollmentResult from GoodServer enroll response
  private JSONObject getEnrollmentResult(JSONObject response) {
    JSONObject enrollmentResult = response.optJSONObject("enrollmentResult");

    if (enrollmentResult == null) {
      return new JSONObject();
    }

    return enrollmentResult;
  }
}
